events/events.go

```diff
@@ -31,9 +35,100 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
+}
+
 // EventRegistrar is an object that you can register EventListeners on.
 type EventRegistrar interface {
-	RegisterListener(EventListener)
+	AddListener(EventListener)
+	RemoveListener(EventListener)
+}
+
+type baseEventRegistrar struct {
+	lock      sync.RWMutex
+	listeners []EventListener
+}
+
+// AddListener adds a listener to the Event Registar. Events emitted on this
+// registar will be invoked on the listener
+func (a *baseEventRegistrar) AddListener(l EventListener) {
+	if l == nil {
+		// do not register nil listener, will cause nil pointer dereference during
+		// event emitting
+		return
+	}
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Check if listener is already registered
+	for _, listener := range a.listeners {
+		if listener == l {
+			return
+		}
+	}
+
+	// by making a copy the backing array will never be changed after its creation.
+	// this allowes to copy the slice while locked but iterate while not locked
+	// preventing deadlocks when listeners are added/removed in the handler of a
+	// listener
+	listenersCopy := make([]EventListener, 0, len(a.listeners)+1)
+	listenersCopy = append(listenersCopy, a.listeners...)
+	listenersCopy = append(listenersCopy, l)
+
+	a.listeners = listenersCopy
+}
+
+// RemoveListener removes a listener from the Event Registar. Subsequent calls
+// to EmitEvent will not cause HandleEvent to be called on this listener.
+func (a *baseEventRegistrar) RemoveListener(l EventListener) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	for i := range a.listeners {
+		if a.listeners[i] == l {
+			// create a new list excluding the listener that needs removal
+			listenersCopy := make([]EventListener, 0, len(a.listeners)-1)
+			listenersCopy = append(listenersCopy, a.listeners[:i]...)
+			listenersCopy = append(listenersCopy, a.listeners[i+1:]...)
+			a.listeners = listenersCopy
+			break
+		}
+	}
+}
+
+// AsyncEventEmitter is an implementation of both an EventRegistar and EventEmitter
+// that emits events in their own go routine.
+type AsyncEventEmitter struct {
+	baseEventRegistrar
+}
+
+// EmitEvent will send the event to all registered listeners
+func (a *AsyncEventEmitter) EmitEvent(event Event) {
+	a.lock.RLock()
+	for _, listener := range a.listeners {
+		go listener.HandleEvent(event)
+	}
+	a.lock.RUnlock()
```

Comment:
We can't omit the lock here because that would facilitate a race on `a.listeners` where listeners are added and removed while events are being emitted. [example](https://travis-ci.org/uber/ringpop-go/jobs/157849594#L530-L576)